Fix unique indexes being added to the UPDATE's WHERE clause while doing an upsert

ke1Del opened the pull request:
Partially fixes #6240 : if UPSERT has PK, only filter UPDATE by PK. Else add the unique indexes logic.

I did not fix the index's where clause not being considered as it seems to me that this logic should not exist. Performing an upsert without specifying any value for the PK seems like a dangerous operation.

PS: I also fixed the Dockerfile that was running `npm install` on the wrong folder and then using the shared node_modules folder. Still `npm run test-docker` is producing a couple of MySQL errors on the v3 branch prior to my changes so I wasn't able to confirm that all tests pass.

janmeier commented:
Please add a test - perhaps just a unit test that stubs out `this.QueryGenerator.upsertQuery` and asserts that its called with the correct `where

ke1Del commented:
Tests were added.
I'm not sure why Travis failed but it doesn't seem related to the changes made.

janmeier commented on docker-compose.yml:
```diff
@@ -3,8 +3,6 @@ sequelize:
   links:
    - mysql
    - postgres
-  volumes:
```
I think volumes is used here to mount the local files so you don't have to rebuild the container each time a file changes - But @mickhansen knows more about that

mickhansen replied:
Should definitely not be removed.

mickhansen commented on Dockerfile:
```diff
@@ -2,8 +2,8 @@ FROM node:5
 
-WORKDIR /sequelize
-VOLUME /sequelize
+COPY . /sequelize/
```
please don't edit dev environment files when you know for a fact that a lot of other people rely on them :)

ke1Del replied:
@mickhansen Sure, I can revert the changes if you want. But I have to warn you that the `.dockerignore` file only applies to the `COPY` operation and not to the `VOLUME`.
Also, you're performing an `npm install` on the `/` (root) folder and then you mount the volume on the `/sequelize` folder, which means the `npm install` is completely useless.
Anyway, the reason I changed it is because I had some package conflicts due to my node_modules folder being compiled with node v6 (which I use on my workstation) and docker_compose using node v5 (I don't remember exactly which module did not like that and needed to be rebuilt).

janmeier replied:
Probably sqlite.

If we can fix the issue of having to have the same node version inside and outside of the container that would be nice :). As long as we can still mount /lib and /test as volumes so we dont need a rebuild when changing files

ke1Del replied:
You can try to create different volumes for each folder but it doesn't look very efficient (you'll probably also need the `.jshintrc` file for the lint test, I don't know about the rest of the files).
A copy operation is actually very fast so rebuilding the container only takes a few seconds. But it's really your choice, I just thought to give a hand as I had some experience with docker containers in the past and it didn't look like the dockerfile was doing what you expected.

mickhansen commented:
Changes to Dockerfile and docker-compose.yml need to be removed completely.

ke1Del commented:
I reverted the changes to the Docker files.
Regarding the failed tests, it doesn't look to me like it's related to the changes I've made. Could you please confirm and let me know if I introduced some kind of regression? Thanks

yahao87 commented:
I need this feature.
When is it officially merged?
